pkg/admission/http.go

```diff
@@ -0,0 +1,301 @@
+package admission
+
+import (
+	"bytes"
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/http/httptest"
+	"strings"
+	"time"
+
+	herrors "github.com/horizoncd/horizon/core/errors"
+	"github.com/horizoncd/horizon/pkg/admission/models"
+	config "github.com/horizoncd/horizon/pkg/config/admission"
+	perror "github.com/horizoncd/horizon/pkg/errors"
+	"github.com/horizoncd/horizon/pkg/util/common"
+)
+
+type HTTPAdmissionClient struct {
+	config config.ClientConfig
+	http.Client
+}
+
+// NewHTTPAdmissionClient creates a new HTTPAdmissionClient
+func NewHTTPAdmissionClient(config config.ClientConfig, timeout time.Duration) *HTTPAdmissionClient {
+	var transport = &http.Transport{}
+	if config.CABundle != "" {
+		ca := config.CABundle
+		certPool := x509.NewCertPool()
+		certPool.AppendCertsFromPEM([]byte(ca))
+		transport = &http.Transport{
+			TLSClientConfig: &tls.Config{
+				RootCAs: certPool,
+			},
+		}
+	}
+	if config.Insecure {
+		transport = &http.Transport{
+			TLSClientConfig: &tls.Config{
+				InsecureSkipVerify: true,
+			},
+		}
+	}
+	return &HTTPAdmissionClient{
+		config: config,
+		Client: http.Client{
+			Timeout:   timeout,
+			Transport: transport,
+		},
+	}
+}
+
+// Get sends the admission request to the webhook server and returns the response
+func (c *HTTPAdmissionClient) Get(ctx context.Context, admitData *Request) (*Response, error) {
+	body, err := json.Marshal(admitData)
+	if err != nil {
+		return nil, err
+	}
+
+	req, err := http.NewRequest("POST", c.config.URL, bytes.NewReader(body))
```

Comment:
Is it suitable to use `NewRequestWithContext` ?

Reply:
You use `ctx, cancelFunc := context.WithCancel(ctx)` in the upper method named `Validating`.

Reply:
fixed